keps/sig-apps/20190715-rollingUpdate-enhancement-for-daemonset.md

```diff
@@ -0,0 +1,181 @@
+
+```go
+// DaemonSetSpec is the specification of a daemon set.
+type DaemonSetSpec struct {
+  ...
+    // Indicates that the daemonset is paused and will not be processed by the
+  // daemonset controller.
+  // +optional
+  Paused bool
+  ...
+}
+```
+
+```go
+// DaemonSetUpdateStrategy is a struct used to control the update strategy for a DaemonSet.
+type DaemonSetUpdateStrategy struct {
+  ...
+	// Surging rolling update config params. Present only if type = "SurgingRollingUpdate".
+	SurgingRollingUpdate *SurgingRollingUpdateDaemonSet `json:"surgingRollingUpdate,omitempty" protobuf:"bytes,3,opt,name=surgingRollingUpdate"`
+}
+```
+
+```go
+// Spec to control the desired behavior of a daemon set surging rolling update.
+type SurgingRollingUpdateDaemonSet struct {
+	// The maximum number of DaemonSet pods that can be scheduled above the desired number of pods
+	// during the update. Value can be an absolute number (ex: 5) or a percentage of the total number
+	// of DaemonSet pods at the start of the update (ex: 10%). The absolute number is calculated from
+	// the percentage by rounding up. This cannot be 0. The default value is 1. Example: when this is
+	// set to 30%, at most 30% of the total number of nodes that should be running the daemon pod
+	// (i.e. status.desiredNumberScheduled) can have 2 pods running at any given time. The update
+	// starts by starting replacements for at most 30% of those DaemonSet pods. Once the new pods are
+	// available it then stops the existing pods before proceeding onto other DaemonSet pods, thus
+	// ensuring that at most 130% of the desired final number of DaemonSet  pods are running at all
+	// times during the update.
+	// +optional
+	MaxSurge *intstr.IntOrString `json:"maxSurge,omitempty" protobuf:"bytes,1,opt,name=maxSurge"`
+
+	// A label query over pods that are managed by the daemon set SurgingRollingUpdate.
+	// Must match in order to be controlled.
+	// It must match the node's labels.
+	Selector *metav1.LabelSelector `json:"selector" protobuf:"bytes,2,opt,name=selector"`
+
+	// The number of DaemonSet pods remained to be old version.
+	// Default value is 0.
+	// Maximum value is status.DesiredNumberScheduled, which means no pod will be updated.
+	// +optional
+	Partition *int32 `json:"partition,omitempty" protobuf:"varint,3,opt,name=partition"`
+}
+```
+
+```go
+type RollingUpdateDaemonSet struct {
+  ...
+	// A label query over nodes that are managed by the daemon set RollingUpdate.
+	// Must match in order to be controlled.
+	// It must match the node's labels.
+	Selector *metav1.LabelSelector `json:"selector" protobuf:"bytes,2,opt,name=selector"`
+
+	// The number of DaemonSet pods remained to be old version.
+	// Default value is 0.
+	// Maximum value is status.DesiredNumberScheduled, which means no pod will be updated.
+	// +optional
+	Partition *int32 `json:"partition,omitempty" protobuf:"varint,3,opt,name=partition"`
```

Comment:
No sure why the existing `MaxUnnavailable` strategy is missing?

Reply:
Sorry for misunderstanding, I  use ”...“ to omit the existed fields.